[OSLogOptimization] Improve the OSLogOptimization pass so that it can fold constant arrays and closure literals

ravikandhadai opened the pull request:
This PR enhances the OSLogOptimization pass so that it can fold constant arrays and closure literals that are inferred by the constant evaluator while evaluating os log calls. The main changes of this PR are mostly confined to the function `emitCodeForSymbolicValue` of the `OSLogOptimization` pass.  In addition to adding support for array and closure folding, this PR also contains a fix to a (minor) bug in the code that manages the lifetime of the folded value.  For perspicuity, this PR is split into 3 self-contained commits as described below: 

 * Commit  9b57208 fixes a bug in the code that replaces a guaranteed value by an owned (constant) value. It also slightly improves the replacement logic for non-OSSA (which is meant to be phased out soon once OSSA is enabled in the mandatory pipeline) so that it can handle the folding of arrays and closures introduced in this PR.  This commit updates `OSLogPrototypeCompileTest.sil` test suite to reflect these changes.

* Commit fa0ec3d adds support for folding constant arrays. It also adds SIL tests for testing array folding logic to the file: `OSLogPrototypeCompileTest.sil`.

* Commit f7b489e adds support for folding symbolic closures, which are representations of closure literals in the constant evaluator. It also adds SIL tests for testing the closure folding logic. Importantly, this commit updates the `OSLogPrototypeCompileTest.swift` to check whether the "array of closures" used by the new OSLog overlay is folded.

ravikandhadai committed:
[OSLogOptimization] Improve the `replaceAndFixLifetimes` function
of the OSLogOptimization pass. This commit contain two changes:
 - It handles non-OSSA better (but it is meant to be phased out) so
   that array and closure folding can be supported
 - It fixes a bug in the OSSA folding by making sure that when an
   owned value replaces a guaranteed value, the owned value is
   borrowed and the borrow is used in place of the guaranteed value.

ravikandhadai committed:
[SIL Optimization][OSLogOptimization] Improve the OSLogOptimization
pass so that it constant folds array symbolic values inferred by the
constant evaluator when evaluting os log calls.

ravikandhadai committed:
[OSLogOptimization] Improve the OSLogOptimization pass so that it can
fold a symbolic closure, which is the representation of a closure literal
in the constant evaluator. This commit improves the function
emitCodeForSymbolicValue so that given a symbolic closure it can emit
SIL code for constructing the closure.

This improvement enables folding the arguments array, which is an array
of closures, by its constant value inferred by constant evaluating the
new OSLog calls.

ravikandhadai commented:
@swift-ci Please test

ravikandhadai commented:
@swift-ci Please test Linux Platform

atrick reviewed (changes requested):
Generally looks good. A couple of comments are inline.

ravikandhadai commented:
@swift-ci Please test

ravikandhadai commented:
@swift-ci Please test

ravikandhadai commented:
@swift-ci Please test

ravikandhadai commented:
@atrick Thanks for the comments. I have pushed a commit that addressed the comments. I have also added test `testNoFoldingOfArrayLiterals` in `OSLogPrototypeCompileTest.sil`  that checks that array literals are not folded.

ravikandhadai commented:
@swift-ci Please test

ravikandhadai commented:
@swift-ci Please test
